src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlColumnEncryptionCngProvider.Windows.cs

```diff
@@ -255,150 +120,93 @@ private void ValidateEncryptionAlgorithm(string encryptionAlgorithm, bool isSyst
             catch (CryptographicException)
             {
-                throw SQL.InvalidCngKey(keyPath, cngProviderName, keyIdentifier, isSystemOp);
+                throw SQL.InvalidCngKey(keyPath, cngProvider.Provider, keyIdentifier, isSystemOp);
             }
-
-            return new RSACng(cngKey);
         }
 
         /// <summary>
-        /// Extracts the CNG provider and key name from the key path
+        /// Extracts the CNG provider and key name from the given key path.
         /// </summary>
-        /// <param name="keyPath">keypath in the format [CNG Provider]/[KeyName]</param>
-        /// <param name="isSystemOp">Indicates if ADO.NET calls or the customer calls the API</param>
-        /// <param name="cngProvider">CNG Provider</param>
-        /// <param name="keyIdentifier">Key identifier inside the CNG provider</param>
-        private void GetCngProviderAndKeyId(string keyPath, bool isSystemOp, out string cngProvider, out string keyIdentifier)
+        /// <param name="keyPath">Key path in the format [CNG provider name]/[key name].</param>
+        /// <param name="isSystemOp">Indicates if ADO.NET calls or the customer calls the API.</param>
+        /// <param name="cngProvider">CNG provider.</param>
+        /// <param name="keyIdentifier">Key name inside the CNG provider.</param>
+        private static void GetCngProviderAndKeyId(string keyPath, bool isSystemOp, out CngProvider cngProvider, out string keyIdentifier)
         {
-            int indexOfSlash = keyPath.IndexOf(@"/", StringComparison.Ordinal);
+            ReadOnlySpan<char> keyPathSpan = keyPath.AsSpan();
+            int indexOfSlash = keyPathSpan.IndexOf('/');
+
             if (indexOfSlash == -1)
             {
                 throw SQL.InvalidCngPath(keyPath, isSystemOp);
             }
-
-            cngProvider = keyPath.Substring(0, indexOfSlash);
-            keyIdentifier = keyPath.Substring(indexOfSlash + 1, keyPath.Length - (indexOfSlash + 1));
-
-            if (cngProvider.Length == 0)
+            else if (indexOfSlash == 0)
             {
                 throw SQL.EmptyCngName(keyPath, isSystemOp);
             }
-
-            if (keyIdentifier.Length == 0)
+            else if (indexOfSlash == keyPath.Length - 1)
             {
                 throw SQL.EmptyCngKeyId(keyPath, isSystemOp);
             }
+
+            ReadOnlySpan<char> cngProviderName = keyPathSpan.Slice(0, indexOfSlash);
+
+            // If the provider is one of the well-known providers, use the static instance instead of allocating a new string and CngProvider.
+            if (cngProviderName.Equals(CngProvider.MicrosoftSoftwareKeyStorageProvider.Provider.AsSpan(), StringComparison.OrdinalIgnoreCase))
+            {
+                cngProvider = CngProvider.MicrosoftSoftwareKeyStorageProvider;
+            }
+            else if (cngProviderName.Equals(CngProvider.MicrosoftSmartCardKeyStorageProvider.Provider.AsSpan(), StringComparison.OrdinalIgnoreCase))
```

Comment:
It seems a bit redundant to repeatedly convert the string to span when there's an overload that directly takes the string. It's just checking for equality so I don't think it will be doing any allocations based on the string.

Reply:
ah, ok, that's fair